### PR TITLE
Fix small bugs in hpopt

### PR DESCRIPTION
**Pull request by akshatzalte**
## Description
Hyperparameter optimization was giving very small (~10^-10) Learning rates, which seemed wrong to people in the devs meeting (5/20/2024). This was because  the search space was too vast.

Also, the batch-size was not enforced to be an integer currently.

Reduction of search space and fixing the values in hpopt.

## Example / Current workflow
In the current workflow, quniform outputs a float value which is not correct. 
For the batch-size, tune.choice() used to simplify it and enforce only 16, 32.., 256 is selected.

## Bugfix / Desired workflow
Reduced the search space for the init and final lr ratios and max_lr Fixed the warmup epochs to be an integer (and not None) Fixed the batch size to be an integer (powers of 2)
Replace quniform to qrandint to ensure int for agrregation_norm, depth, ffn_hidden_dim, ffn_num_layers, message_hidden_dim.

## Questions
none

## Relevant issues
#851

### CODE REVIEW

File: chemprop/cli/hpopt.py [hunk 1]
```diff
@@ -43,16 +43,16 @@
     DEFAULT_SEARCH_SPACE = {
         "activation": tune.choice(categories=list(Activation.keys())),
         "aggregation": tune.choice(categories=list(AggregationRegistry.keys())),
-        "aggregation_norm": tune.quniform(lower=1, upper=200, q=1),
-        "batch_size": tune.loguniform(lower=16, upper=256, base=2),
-        "depth": tune.quniform(lower=2, upper=6, q=1),
+        "aggregation_norm": tune.qrandint(lower=1, upper=200, q=1),
```
Comment (KnathanM) [inline chemprop/cli/hpopt.py:46]: Why did you change this to an integer? NormAggregation expects the norm to be a float. Also looks like the `q=1` isn't needed for this and the other `qrandint`s.

File: chemprop/cli/hpopt.py [hunk 2]
```diff
@@ -43,16 +43,16 @@
...
+        "batch_size": tune.choice(np.logspace(start=4, stop=8, num=5, base=2)),
```
Comment (KnathanM) [inline chemprop/cli/hpopt.py:47]:
```suggestion
        "batch_size": tune.choice(np.logspace(start=4, stop=8, num=5, base=2, dtype=int)),
```

Then I don't think you need to set it to an int down in update_args_with_config. I could be wrong about that if the number is still read as a float from the config file

File: chemprop/cli/hpopt.py [hunk 3]
```diff
@@ -225,7 +225,7 @@ def process_hpopt_args(args: Namespace) -> Namespace:
 
 
 def build_search_space(search_parameters: list[str], train_epochs: int) -> dict:
-    if "warmup_epochs" not in SEARCH_SPACE and "warmup_epochs" in search_parameters:
+    if "warmup_epochs" in search_parameters:
         SEARCH_SPACE["warmup_epochs"] = tune.quniform(lower=1, upper=train_epochs // 2, q=1)
```
Comment (KnathanM) [inline chemprop/cli/hpopt.py:229]: Should this be `qrandint` as well?

**Review by KnathanM — COMMENTED** [review]
Thanks for the PR!

File: chemprop/cli/hpopt.py (hunk 2 above)
Comment (JacksonBurns) [inline chemprop/cli/hpopt.py:47]: Can this be rewritten using `tune`'s [`qlograndint`](https://docs.ray.io/en/latest/tune/api/doc/ray.tune.qlograndint.html#ray-tune-qlograndint) method?

File: chemprop/cli/hpopt.py [hunk 4]
```diff
@@ -242,6 +242,9 @@ def update_args_with_config(args: Namespace, config: dict) -> Namespace:
             case "init_lr_ratio":
                 setattr(args, "init_lr", value * args.max_lr)
 
+            case "batch_size":
+                setattr(args, "batch_size", int(value))
```
Comment (JacksonBurns) [inline chemprop/cli/hpopt.py:246]: Whatever `tune` method we call should return an `int`, making this cast unnecessary?

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
Couple small questions/suggestions. Thanks for starting this!

File: chemprop/cli/hpopt.py (hunk 4 above)
Comment (JacksonBurns) [reply chemprop/cli/hpopt.py:246]: Also mentioned here by @KnathanM: https://github.com/chemprop/chemprop/pull/873#discussion_r1608761150

File: chemprop/cli/hpopt.py [hunk 5]
```diff
@@ -225,7 +225,7 @@ def process_hpopt_args(args: Namespace) -> Namespace:
 
 
 def build_search_space(search_parameters: list[str], train_epochs: int) -> dict:
-    if "warmup_epochs" not in SEARCH_SPACE and "warmup_epochs" in search_parameters:
```
Comment (shihchengli) [inline chemprop/cli/hpopt.py:228]: Why remove `warmup_epochs" not in SEARCH_SPACE`? If one uses a self-defined search space, it would rewrite the search range to `tune.quniform(lower=1, upper=train_epochs // 2, q=1)`.

**Review by shihchengli — COMMENTED** [review]
Thanks for making this PR.

File: chemprop/cli/hpopt.py (hunk 1 above)
Comment (akshatzalte) [reply chemprop/cli/hpopt.py:46]: I will change it back to quniform()
Also, q = 1 is the default, so maybe we can just leave it as it is.

File: chemprop/cli/hpopt.py (hunk 2 above)
Comment (akshatzalte) [reply chemprop/cli/hpopt.py:47]: Good catch!
Implementing this.

File: chemprop/cli/hpopt.py (hunk 3 above)
Comment (akshatzalte) [reply chemprop/cli/hpopt.py:229]: Yes, changing that

File: chemprop/cli/hpopt.py (hunk 2 above)
Comment (akshatzalte) [reply chemprop/cli/hpopt.py:47]: I tried that but it did not exclusively sample powers of 2. It is probably just doing a log distribution over all integers.

File: chemprop/cli/hpopt.py (hunk 4 above)
Comment (akshatzalte) [reply chemprop/cli/hpopt.py:246]: Yes, good catch.
Working on it

File: chemprop/cli/hpopt.py (hunk 5 above)
Comment (akshatzalte) [reply chemprop/cli/hpopt.py:228]: Good catch. We will rewrite it.
We removed it as initially it was always returning "None" (which was its initialization for default search space dict) for each case.

**Comment by akshatzalte** [comment]
I have addressed all the comments. Please re-review and merge!

File: chemprop/cli/hpopt.py (hunk 2 above)
Comment (akshatzalte) [reply chemprop/cli/hpopt.py]: Here is an example run:
`>>> from ray import tune
>>> sampler = tune.qlograndint(lower=2,upper=8,base=2,q=1)
>>> sampler.sample()
2
>>> sampler.sample()
4
>>> sampler.sample()
7`

**Comment by akshatzalte** [comment]
Made some more fixes re: warmup epochs. Open for re-review.

File: chemprop/cli/hpopt.py (hunk 4 above)
Comment (shihchengli) [reply chemprop/cli/hpopt.py:248]: @akshatzalte I think this hasn't been addressed yet. As `tune.choice(np.logspace(start=4, stop=8, num=5, base=2, dtype=int))` is used, `batch_size` should always be an integer. The code here can be removed.

File: chemprop/cli/hpopt.py [hunk 6]
```diff
@@ -46,15 +46,15 @@
         "activation": tune.choice(categories=list(Activation.keys())),
         "aggregation": tune.choice(categories=list(AggregationRegistry.keys())),
         "aggregation_norm": tune.quniform(lower=1, upper=200, q=1),
-        "batch_size": tune.loguniform(lower=16, upper=256, base=2),
-        "depth": tune.quniform(lower=2, upper=6, q=1),
+        "batch_size": tune.choice(np.logspace(start=4, stop=8, num=5, base=2, dtype=int)),
```
Comment (shihchengli) [inline chemprop/cli/hpopt.py:49]: I am thinking about just using `tune.choice([16, 32, 64, 128, 256])`

Comment (akshatzalte) [reply chemprop/cli/hpopt.py:49]: Yes, the current line is equivalent. logspace() outputs the [16, 32, 64, 128, 256] vector.
Let me know if you still think its not equivalent.

Comment (shihchengli) [reply chemprop/cli/hpopt.py:49]: I meant replace `np.logspace(start=4, stop=8, num=5, base=2, dtype=int)` with `[16, 32, 64, 128, 256]`. I think it is clearer and there is no downside to using this.

Comment (akshatzalte) [reply chemprop/cli/hpopt.py:49]: I think the current implementation is better

Comment (KnathanM) [reply chemprop/cli/hpopt.py:49]: I agree with Shih-Cheng that `[16, 32, 64, 128, 256]` is better than using `np.logspace` because it is immediately obvious what the choices are. With `np.logspace` a code reader has to either run the line of code (as I did when reviewing this before) or know what `np.logspace` does and their powers of two. 

I could see the current implementation being better if we expect to change the choices significantly in the future as it is more flexible, but that flexibility probably isn't need.

**Review by shihchengli — APPROVED** [review]
LGTM

File: chemprop/cli/hpopt.py [hunk 7]
```diff
@@ -227,8 +226,8 @@ def process_hpopt_args(args: Namespace) -> Namespace:
 
 
 def build_search_space(search_parameters: list[str], train_epochs: int) -> dict:
-    if "warmup_epochs" not in SEARCH_SPACE and "warmup_epochs" in search_parameters:
-        SEARCH_SPACE["warmup_epochs"] = tune.quniform(lower=1, upper=train_epochs // 2, q=1)
+    if "warmup_epochs" in search_parameters and SEARCH_SPACE.get("warmup_epochs", None) is None:
```
Comment (JacksonBurns) [inline chemprop/cli/hpopt.py:229]:
```suggestion
    if "warmup_epochs" in search_parameters and SEARCH_SPACE.get("warmup_epochs", True):
```

File: chemprop/cli/hpopt.py (hunk 2 above)
Comment (JacksonBurns) [reply chemprop/cli/hpopt.py]: I like how this turned out later on

**Comment by JacksonBurns** [comment]
@akshatzalte one last small thing, then it's good to go.

File: chemprop/cli/hpopt.py (hunk 7 above)
Comment (JacksonBurns) [reply chemprop/cli/hpopt.py:229]: Ah good point! Not an equivalent suggestion

**Review by JacksonBurns — APPROVED** [review]

File: chemprop/cli/hpopt.py (hunk 7 above)
Comment (akshatzalte) [reply chemprop/cli/hpopt.py:229]: Can you explain what happens when a custom range is provided  by the user? The `SEARCH_SPACE.get("warmup_epochs", True)` will evaluate to that range which will either give a syntax error while using with `if` or it may evaluate to `True` as it is a non-zero number which is not what we want. Please let me know! I'll try running some cases to better my understanding.

Comment (JacksonBurns) [reply chemprop/cli/hpopt.py:229]: It would result in True since a non-empty list is True, but don't worry about it, I merged without this suggestion based on your comment above.
